system_tests/recreatestate_rpc_test.go

```diff
@@ -259,9 +262,166 @@ func TestRecreateStateForRPCBlockNotFoundWhileRecreating(t *testing.T) {
+	}
+	expectedBalance, err := client.BalanceAt(ctx, GetTestAddressForAccountName(t, "User2"), new(big.Int).SetUint64(lastBlock))
+	Require(t, err)
+
+	node.StopAndWait()
+	cancel1()
+	t.Log("stopped first node")
+
+	AddDefaultValNode(t, ctx, nodeConfig, true)
+	l2info, stack, chainDb, arbDb, blockchain = createL2BlockChain(t, l2info, dataDir, params.ArbitrumDevTestChainConfig(), cacheConfig)
+	node, err = arbnode.CreateNode(ctx, stack, chainDb, arbDb, NewFetcherFromConfig(nodeConfig), blockchain, nil, node.DeployInfo, nil, nil, nil, feedErrChan)
+	Require(t, err)
+	Require(t, node.Start(ctx))
+	client = ClientForStack(t, stack)
+	defer node.StopAndWait()
+	bc = node.Execution.Backend.ArbInterface().BlockChain()
+	gas := skipGas
+	blocks := skipBlocks
+	for i := genesis + 1; i <= genesis+uint64(txCount); i++ {
+		block := bc.GetBlockByNumber(i)
+		if block == nil {
+			Fatal(t, "header not found for block number:", i)
+			continue
+		}
+		gas += block.GasUsed()
+		blocks++
+		_, err := bc.StateAt(block.Root())
+		if (skipBlocks == 0 && skipGas == 0) || (skipBlocks != 0 && blocks > skipBlocks) || (skipGas != 0 && gas > skipGas) {
+			if err != nil {
+				t.Log("blocks:", blocks, "skipBlocks:", skipBlocks, "gas:", gas, "skipGas:", skipGas)
+			}
+			Require(t, err, "state not found, root:", block.Root(), "blockNumber:", i, "blockHash", block.Hash(), "err:", err)
+			gas = 0
+			blocks = 0
+		} else {
+			if err == nil {
+				t.Log("blocks:", blocks, "skipBlocks:", skipBlocks, "gas:", gas, "skipGas:", skipGas)
+				Fatal(t, "state shouldn't be available, root:", block.Root(), "blockNumber:", i, "blockHash", block.Hash())
+			}
+			expectedErr := &trie.MissingNodeError{}
+			if !errors.As(err, &expectedErr) {
+				Fatal(t, "getting state failed with unexpected error, root:", block.Root(), "blockNumber:", i, "blockHash", block.Hash())
+			}
+		}
+	}
+	for i := genesis + 1; i <= genesis+uint64(txCount); i += i % 10 {
+		_, err = client.BalanceAt(ctx, GetTestAddressForAccountName(t, "User2"), new(big.Int).SetUint64(i))
+		Require(t, err)
+	}
+
+	balance, err := client.BalanceAt(ctx, GetTestAddressForAccountName(t, "User2"), new(big.Int).SetUint64(lastBlock))
+	Require(t, err)
+	if balance.Cmp(expectedBalance) != 0 {
+		Fatal(t, "unexpected balance result for last block, want: ", expectedBalance, " have: ", balance)
+	}
+}
+
+func TestSkippingSavingStateAndRecreatingAfterRestart(t *testing.T) {
+	// test defaults
+	testSkippingSavingStateAndRecreatingAfterRestart(t, 0, 0, 512)
+	testSkippingSavingStateAndRecreatingAfterRestart(t, 127, 0, 512)
+	testSkippingSavingStateAndRecreatingAfterRestart(t, 0, 15*1000*1000, 512)
+	testSkippingSavingStateAndRecreatingAfterRestart(t, 127, 15*1000*1000, 512)
+
+	// one test block ~ 925000 gas
+	testBlockGas := uint64(925000)
+	skipGasValues := []uint64{testBlockGas, 2 * testBlockGas, 3 * testBlockGas, 5 * testBlockGas, 21 * testBlockGas}
+	skipBlockValues := []uint32{1, 2, 3, 5, 21}
```

Comment:
I'd start both these lists with 0 (and remove the first call to testSkipping(t,0,0,512)